languages/source.nim

```diff
@@ -291,6 +291,28 @@ const lang* = language:
     of type(typ): true
     of typ:       false
 
+  func unionOfAux(list: *typ, accum: typ) -> typ =
+    # `TupleTy` is used as the accumulator, because it can hold any number
+    # of types
+    case list
+    of [typ_1, *typ_2]:
+      if typ_1 in typ_2:
+        # it's a duplicate; skip
+        unionOfAux(typ_2, accum)
+      else:
+        case accum
+        of TupleTy(*typ_3):
+          unionOfAux(typ_2, TupleTy(...typ_3, typ_1))
+    else:
+      accum
+
+  func unionOf(list: *typ) -> typ =
```

Comment:
Ah, so this "flattens the union", or canonicalizes/normalizes it, meaning an operation resulting in the smallest type representation that still matches the same types as the previous one.

It _might_ be worth renaming to `normalizeUnion`, or derivative, but I'm not convinced and/or don't feel too strongly about it.

Reply:
> Ah, so this "flattens the union", or canonicalizes/normalizes it, meaning an operation resulting in the smallest type representation that still matches the same types as the previous one.

Hmm, not quite. I think a more accurate description would be "treat this unordered *list* of types as a *set* (meaning no duplicates) and create a `UnionTy` from it". So it's not `union -> union`, but rather `bag of types -> maybe union`.

I'd have preferred the "smallest type" operation, but that'd be problematic for multiple reasons (one of them outlined [here](https://github.com/nim-works/phy/pull/135#issuecomment-2746413647)).